milestones/data.py

```diff
@@ -0,0 +1,390 @@
+    if relationship is not None:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = queryset.filter(
+            milestone_relationship_type=mrt.id,
+        )
+
+    # To pull the list of milestones a user HAS, use get_user_milestones
+    # Use fetch_courses_milestones to pull the list of milestones that a user does not yet
+    # have for the specified course
+    if relationship == RELATIONSHIP_TYPES['REQUIRES'] and user and user.get('id', 0) > 0:
+        queryset = queryset.exclude(milestone__usermilestone__user_id=user['id'])
+
+    # Assemble the response container
+    course_milestones = []
+    if len(queryset):
+        for milestone in queryset:
+            course_milestones.append(serializers.serialize_milestone_with_course(milestone))
+
+    return course_milestones
+
+
+def create_course_content_milestone(course_key, content_key, relationship, milestone):
+    """
+    Inserts a new course-content-milestone into app/local state
+    No response currently defined for this operation
+    """
+    relationship_type = _get_milestone_relationship_type(relationship)
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    internal.CourseContentMilestone.objects.get_or_create(
+        course_id=unicode(course_key),
+        content_id=unicode(content_key),
+        milestone=milestone_obj,
+        milestone_relationship_type=relationship_type,
+        active=True,
+    )
+
+
+def delete_course_content_milestone(course_key, content_key, milestone):
+    """
+    Removes an existing course-content-milestone from app/local state
+    No response currently defined for this operation
+    """
+    try:
+        internal.CourseContentMilestone.objects.get(
+            course_id=unicode(course_key),
+            content_id=unicode(content_key),
+            milestone=milestone['id'],
+            active=True,
+        ).delete()
+    except internal.CourseContentMilestone.DoesNotExist:
+        pass
+
+
+def fetch_course_content_milestones(content_key, course_key=None, relationship=None):
+    """
+    Retrieves the set of milestones currently linked to the specified course content
+    Optionally pass in 'relationship' (ex. 'fulfills') to filter down the set
+    Optionally pass in 'user' to further-filter the set (ex. for retrieving unfulfilled milestones)
+    """
+    queryset = internal.CourseContentMilestone.objects.filter(
+        active=True
+    ).select_related('milestone')
+
+    if course_key is not None:
+        queryset = queryset.filter(course_id=unicode(course_key)).select_related('milestone')
+
+    if content_key is not None:
+        queryset = queryset.filter(content_id=unicode(content_key)).select_related('milestone')
+
+    if relationship is not None:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = queryset.filter(milestone_relationship_type=mrt.id).select_related('milestone')
```

Comment:
I don't think you have to specify `select_related` more than once at the top here.